**Xing Yi Chan**

**R00183768**

## **Part 1**
The first task is to choose from two natural language statements with similar wordings which one makes sense and which one does not make sense.

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

### **Preparing Training Data**

In [ ]:
# import both training data and training labels
train_data = pd.read_csv('/content/drive/My Drive/NLP/dataset2/traindata/subtaskA_data_all.csv')
train_label = pd.read_csv('/content/drive/My Drive/NLP/dataset2/traindata/subtaskA_answers_all.csv', names=['id', 'ans'])

# extract positive sentences
train_pos1 = train_data[train_label['ans'] == 0]['sent1']
train_pos2 = train_data[train_label['ans'] == 1]['sent0']
train_pos_sent = train_pos1.append(train_pos2)

# extract negative sentences
train_neg1 = train_data[train_label['ans'] == 0]['sent0']
train_neg2 = train_data[train_label['ans'] == 1]['sent1']
train_neg_sent = train_neg1.append(train_neg2)

#combining both positive and negative sentences
   # sensible sentences --> 0
   # nonsense sentences --> 1
trainX = train_pos_sent.append(train_neg_sent)
# create labels for the sentences
trainY = pd.Series([0]*10000).append(pd.Series([1]*10000))

### **Preparing Testing Data**

In [ ]:
# import both training data and training labels
test_data = pd.read_csv('/content/drive/My Drive/NLP/dataset2/testdata/subtaskA_trial_data.csv')
test_label = pd.read_csv('/content/drive/My Drive/NLP/dataset2/testdata/subtaskA_answers.csv', names=['id', 'ans'])

# extract positive sentences
test_pos1 = test_data[test_label['ans'] == 0]['sent1']
test_pos2 = test_data[test_label['ans'] == 1]['sent0']
train_pos_sent = test_pos1.append(test_pos2)

# extract negative sentences
test_neg1 = test_data[test_label['ans'] == 0]['sent0']
test_neg2 = test_data[test_label['ans'] == 1]['sent1']
test_neg_sent = test_neg1.append(test_neg2)

#combining both positive and negative sentences
   # sensible sentences --> 0
   # nonsense sentences --> 1
testX = train_pos_sent.append(test_neg_sent)
# create labels for the sentences
testY = pd.Series([0]*2021).append(pd.Series([1]*2021))

## **Create tf-idf vectors for both training and testing data**

In [ ]:
# transform words to integers using TfidfVectorizer() function
vectorizer = TfidfVectorizer(max_features=2000, analyzer='word', ngram_range=(1, 4))
trainX_tfidf = vectorizer.fit_transform(trainX)
testX_tfidf = vectorizer.fit_transform(testX)

## **Performs classification to get accuracy score**

In [ ]:
# naive bayes classifier
nb = MultinomialNB()
nb.fit(trainX_tfidf, trainY)
print('Accuracy of naive bayes: ', nb.score(testX_tfidf, testY)*100)

# k-nearest neighbour classifier     
knn = KNeighborsClassifier()
knn.fit(trainX_tfidf, trainY)
print('Accuracy of k-nn: ', knn.score(testX_tfidf, testY)*100)

# random forest classifier
rf = RandomForestClassifier()
rf.fit(trainX_tfidf, trainY)
print('Accuracy of random forest: ', rf.score(testX_tfidf, testY)*100)

# svm classifier
svm = SVC()
svm.fit(trainX_tfidf, trainY)
print('Accuracy of svm: ', svm.score(testX_tfidf, testY)*100)

Accuracy of naive bayes:  51.53389411182583
Accuracy of k-nn:  48.98565066798615
Accuracy of random forest:  50.643245917862444
Accuracy of svm:  50.569025235032164


### **Conclusion**

For classification, 4 different type of classification methods had been used. They include Multinomial Naive Bayes Classifier, k-Nearest Neighbour Classifier, Random Forest Classifier and Support Vector Machine Classifier. After comparing the accuracy of these classifiers, it can be concluded that all these classifiers will give an accuracy score between 50%. 

Among all the 4 classifiers, Multinomial Naive Bayes classifier performs best with the accuracy of 51.53%.